### DISCUSSION TITLE
async validation

### DISCUSSION THREAD

**Discussion by dfa1** [graphql-java/graphql-java#3296] [General]
In one `Instrumentation` I need to invoke an external service  
as part of `beginValidation`. The external service returns another `CompletableFuture` but I don't see a way to combine it with `InstrumentationContext` callbacks.

Any suggestion is welcome!

PS
Maybe the context callbacks should have `CompletableFuture`, in order to allow chaining?

**Comment by bbakerman** [comment]
There is no happens before relationship between Instrumentation calls and the main engine running

The engine tells `Instrumentation` that it started (beginXXX) and that it completed that stage via the `InstrumentationContext` callback but it does not have to wait for that Instrumentation


Some one also asked for this in another way : https://github.com/graphql-java/graphql-java/issues/3291

We are very hesitant to add a "happens before" relationship between `Instrumentation` and the main engine - that is the engine waiting for a chained CF coming back from the `Instrumentation`


You DO get the CompletableFuture during initial dispatch

```
void onDispatched(CompletableFuture<T> result);
```

My question to you is - what do you want to do here?

You make a call to some external system during beginValidation - are you trying to actually do validation during this call?

Because `Instrumentation.beginValidation` is advistory - its started and it will tell you when its finished BUT the `Instrumentation` is not meant to be able to influence the validation result by design

Perhaps `graphql.execution.preparsed.PreparsedDocumentProvider` could help you - its nominally provides the document from say an external store but `parseAndValidateFunction` allows you to return errors becauses its a nominally callback to do actual parsing and validation

You could call that and then wrap it with extra validation from some external service.


```
default CompletableFuture<PreparsedDocumentEntry> getDocumentAsync(ExecutionInput executionInput, Function<ExecutionInput, PreparsedDocumentEntry> parseAndValidateFunction) {
```

**Reply by dfa1** [reply]
thanks @bbakerman for the reply!

Some context: I'm using `Instrumentation` to build an authorization layer. The external service call is needed to get what the user can actually access in the graph and it is used to fill up the `InstrumentationState`.

I could not put this data in `createState` because it is not returning a CF, so `beginValidation` was used but the callback is blocking. 

What if we change `createState` to be CF? That would solve my problem and perhaps it is symmetrical with the CF for `instrumentExecutionResult` (that I cannot use because it is happening too late).

Anyway, in the meantime I'm going to use the `PreparedDocumentProvider`... Thanks for the suggestion!

**Reply by bbakerman** [reply]
> to build an authorization layer. 
Whats the preconditions here

Should the auth information been know before ANY graphql execution is performed

nominally it feels like something that should be established BEFORE the engine even gets started and passed in perhaps via GraphqlContext

But I dont know what you want to auth so its hard to know.

**Reply by dfa1** [reply]
Yes, the idea was to keep all the necessary auth logic in one place (i.e. the `Instrumentation`). 

Perhaps I should propose a PR to make `createState` with CF? Are you willing to review that?

**Comment by bbakerman** [comment]
Sure

**Comment by bbakerman** [comment]
> Perhaps I should propose a PR to make createState with CF?

Thinking on this more I dont see it as a problem - the `Instrumentation.createState()` call is one of the first calls made inside `public CompletableFuture<ExecutionResult> executeAsync(ExecutionInput executionInput)`  - it can become a composed CF call.

It does makes sense to have `Instrumentation.createState()` as async since it is a starting mechanism and not an advistory mechanism like the beginXXX / onComplete calls

If a graphql `instrumentation` chooses to take a long time to get started (eg call other services) - so be it - its on that implementation.

We will want to default backwards to the old createState for compatibility reasons

```
    default CompleteableFuture<InstrumentationState> createStateAsync(InstrumentationCreateStateParameters parameters) {
        return CompleteableFuture.completedFuture(createState());
    }
```

The tricky part will be in `graphql.execution.instrumentation.ChainedInstrumentation.ChainedInstrumentationState` where we need to combine N Cfs together and wait for them all in an async manner

**Reply by dfa1** [reply]
yes, exactly my idea to preserve backward compatibility 😇

Can we use Async sequentially utility method to compose `createState` CFs?

D.

**Reply by bbakerman** [reply]
See https://github.com/graphql-java/graphql-java/pull/3302

**Reply by dfa1** [reply]
@bbakerman it looks good, thanks! ☺️Would it be possible to merge it to master? I could do some early testing on a project before the release of graphql-java!

**Comment by dfa1** [comment]
@bbakerman it is working as expected! thank you very much for helping here!
